src/System.Buffers/src/System/Buffers/DefaultArrayPool.cs

```diff
@@ -0,0 +1,151 @@
+            _rentedBufferCount = 0;
+            _totalBufferCount = 0;
+        }
+
+        public override T[] Rent(int length)
+        {
+            if (length <= 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+
+            Interlocked.Increment(ref _rentedBufferCount);
+
+            int index = Utilities.SelectBucketIndex(length);
+            if (index >= _buckets.Length)
+            {
+                Interlocked.Increment(ref _totalBufferCount);
+                return new T[length];
+            }
+            else
+            {
+                T[] buffer = null;
+
+                // Search for an array starting at the 'index' bucket. If the bucket
+                // is empty, bump up to the next higher bucket and try that one
+                for (int i = index; i < _buckets.Length; i++)
+                {
+                    // Check if the bucket has been created and, if not, create it
+                    ArrayPoolBucket<T> bucket = Volatile.Read(ref _buckets[i]);
+                    if (bucket == null)
+                    {
+                        if (Interlocked.CompareExchange(ref _buckets[index], new ArrayPoolBucket<T>(Utilities.GetMaxSizeForBucket(index), _numberOfArraysPerBucket), null) == null)
+                        {
+                            Interlocked.Add(ref _totalBufferCount, _numberOfArraysPerBucket);
+                        }
+                        bucket = _buckets[index];
+                    }
+
+                    buffer = bucket.Rent();
+
+                    // If the bucket has an array left and returned it, give it to the caller
+                    if (buffer != null)
+                    {
+                        return buffer;
+                    }
+                }
+
+                // Getting here means that all Buckets from index through _buckets.Length are taken
+                // so we need to allocate a buffer on demand
+                Interlocked.Increment(ref _totalBufferCount);
+                return new T[length];
+            }
+        }
+
+        public override T[] Enlarge(T[] original, int newLength)
+        {
+            if (original == null)
+                throw new ArgumentNullException(nameof(original));
+            // Our buffers have a minimum size of 16, so check for that first
+            if ((newLength > MinimiumArraySize || original.Length > MinimiumArraySize) &&
+                (newLength < original.Length))
+                throw new ArgumentOutOfRangeException(nameof(newLength), SR.Invalid_Enlarge_New_Size);
+
+            // Since we can't realloc, get a bigger buffer, copy the data, clear the old buffer, and return the new one
+            T[] newBuffer = Rent(newLength);
+            Array.Copy(original, newBuffer, original.Length);
+            Return(original, clearArray: true);
+
+            return newBuffer;
+        }
+
+        public override void Return(T[] buffer, bool clearArray = false)
```

Comment:
Since we sometimes ignore the value of clearArray and decide to always clear, should it be ```forceClearArray``` or something like that which indicates we may choose to clear even if it wasn't asked for?

Or maybe the whole notion of clearing automatically should be removed, and if the developer wants to clear, they can call Array.Clear themselves before calling Return?